File: leet-code/234. Palindrome Linkded List/PalindromeLinkedList234_3.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

import java.util.*;


public class PalindromeLinkedList234_3 {
    public boolean isPalindrome(ListNode head) {
        ArrayDeque<Integer> deque = new ArrayDeque<>();
        ListNode cur = head;
        while (cur != null) {
            deque.add(cur.val);
            cur = cur.next;
        }
        while (!deque.isEmpty() && deque.size() > 1) {
            if (!Objects.equals(deque.pollFirst(), deque.pollLast())) {
                return false;
            }
        }
        return true;
    }
}

File: leet-code/234. Palindrome Linkded List/PalindromeLinkedList234_3.kt
package com.example

class PalindromeLinkedList234_3 {
    fun isPalindrome(head: ListNode?): Boolean {
        var prev: ListNode? = null
        var cur: ListNode? = head
        var next: ListNode? = head
        while (next != null && next.next != null) {
            prev = cur
            cur = cur?.next
            next = next.next.next
        }
        prev?.next = null
        if (next != null) {
            cur = cur?.next
        }

        var front: ListNode? = null
        while (cur != null) {
            val next = cur.next
            cur.next = front
            front = cur
            cur = next
        }

        var node: ListNode? = head
        while (node != null && front != null && node.`val` == front.`val`) {
            node = node.next
            front = front.next
        }
        return front == null
    }
}